### PR TITLE
JDK-8236048: Cleanup use of Utils.normalizeNewlines

### PR DESCRIPTION
**Pull request by jonathan-gibbons**
Please review a medium-size cleanup related to handling newlines in `Content`/`HtmlTree`/`Text[Builder]` trees.
This is the 3rd and last in the recent series of cleanup tasks in the code to generate HTML.

Until recently, the `Text`/`TextBuilder` objects were "ready-for-writing", meaning that characters like `<`, `&`, etc were encoded as entities, and newlines were represented by the platform line separator. This caused difficulties when counting the size of generated text, since these characters had to be detected and accounted for.  A recent change addressed entities, but stayed clear of the newline issue, which is now addressed here.

A related performance problem is that the method `Utils.normalizeNewlines` always created new strings, even when it was not necessary.

With this change, newlines in `Text`, `TextBuilder` and other subtypes of `Content` containing text are always represented by `\n`, and are converted to the platform line separator as needed. 

I note the following assumptions:

* the input may contain any flavor of newlines in the user-provided documentation comments
* the output should aways use the platform line separator (`\n` on Linux and macOS, `\r\n` on Windows etc.)

Together, these imply that some amount of the input will need to be scanned to normalize newlines at some point in the process. While it would be nice to not have to normalize newlines, it's a case or "pay now, or pay later". 

Notable parts of this change:

* `Utils.normalizeNewlines` is moved to be a static method in `Text` and rewritten to not create new strings unless needed. This allows a number of questionable imports of `toolkit.utils.DocletConstants` to be removed from the `html.markup` package.  While it would be possible to call this method automatically on all strings passed to `Text`, `TextBuilder` etc, in many cases it is not necessary: put another way, it only needs to be called on strings that have come from the user, either in documentation comments or in command-line options.
* `Content.write`, and the implementations thereof, are given a new parameter which gives the newline sequence to use. This is set to `\n` for `toString` and the platform separator when writing to a file.
* `DocletConstants.NL` goes away and is superseded by:
   *  `Text.NL`: always `\n`, to be used when adding text to a `Content` tree (this replaces most uses of `DocletConstants.NL`)
   * `DocFile.PLATFORM_LINE_SEPARATOR`: the platform line separator, to be used when writing to files
* The various `charCount` methods no longer have to worry about use of the platform line separator 
* Assertions are used to verify that there are no `\r` characters added into `Text`/`TextBuilder`/`RawHtml`/`Content`

Other cleanup:
* `DocletConstants.DEFAULT_TAB_STOP_LENGTH` is moved to `BaseOptions` which is the only class that uses it. This leaves just two remaining constants in `DocletConstants` which should probably be moved elsewhere as well, and the `DocletConstants` class deleted. That is a step too far for the work here.
* Minor IDE-suggestions, for lambdas, enhanced switch, adding braces, etc

One test was affected by the `DocletConstants.NL` change, but apart from that, this is a pure-cleanup change, with no (intentional) externally visible effect.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8236048](https://bugs.openjdk.org/browse/JDK-8236048): Cleanup use of Utils.normalizeNewlines

### Reviewers
 * [Hannes Wallnöfer](https://openjdk.org/census#hannesw) (@hns - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9691/head:pull/9691` \
`$ git checkout pull/9691`

Update a local copy of the PR: \
`$ git checkout pull/9691` \
`$ git pull https://git.openjdk.org/jdk pull/9691/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9691`

View PR using the GUI difftool: \
`$ git pr show -t 9691`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9691.diff">https://git.openjdk.org/jdk/pull/9691.diff</a>

</details>

### CODE REVIEW

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/markup/HtmlDocument.java [hunk 1]
```diff
@@ -68,16 +67,16 @@ public void write(DocFile docFile) throws DocFileIOException {
...
         }
     }
 
-    private void write(Writer writer) throws IOException {
+    private void write(Writer writer, String newline) throws IOException {
         writer.write(docType.text);
-        writer.write(DocletConstants.NL);
-        docContent.write(writer, true);
+        writer.write(newline);
+        docContent.write(writer, newline,true);
```
Comment (hns) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/markup/HtmlDocument.java:80]: Missing space after comma

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/markup/Text.java [hunk 2]
```diff
@@ -47,6 +48,7 @@ public class Text extends Content {
      * @return the object
      */
     public static Text of(CharSequence content) {
+        assert checkNewlines(content);
```
Comment (hns) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/markup/Text.java:51]: Redundant assert as private constructor also contains it.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/markup/RawHtml.java [hunk 3]
```diff
@@ -151,14 +151,6 @@ protected static int charCount(CharSequence htmlText) {
...
                             count++;
                             break;
-                        case '\r':
-                            // Windows uses "\r\n" as line separator while UNIX uses "\n".
-                            // Skip the "\r" to get consistent results across platforms.
-                            if (i + 1 < htmlText.length() && htmlText.charAt(i + 1) == '\n') {
-                                i++;
-                            }
-                            count++;
-                            break;
```
Comment (hns) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/markup/RawHtml.java:161]: The RawHtml constructor contains `assert Text.checkNewlines(rawHtml)` so I'm not sure this is actually needed.

**Review by hns — COMMENTED** [review]
This looks quite good. Apart from the inline comments, I wondered about the following:

 - Would the methods and constant you moved to `Text` be better  placed in `Content` as they are used in other Content instances?
 - There are quite a few uses of the `"\n"` string literal in a content or pre-content context where the `NL` constant couuld be used.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/markup/HtmlDocument.java (hunk 1 above)
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/markup/HtmlDocument.java:80]: Fixed.  
This is a mildly annoying side-effect of IDEA wanting to label boolean arguments: it makes the missing spaces much harder to see.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/markup/Text.java (hunk 2 above)
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/markup/Text.java:51]: Fixed

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/markup/RawHtml.java (hunk 3 above)
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/markup/RawHtml.java:161]: I'm not sure what you are referring to here. Yes, there is the assertion about newlines, but you still have to scan the `rawHtml` for start and end elements, and entities.  Note there is no longer any special handling of newline characters (`\r`, `\n` etc)

**Comment by jonathan-gibbons** [comment]
> This looks quite good. Apart from the inline comments, I wondered about the following:
> 
> * Would the methods and constant you moved to `Text` be better  placed in `Content` as they are used in other Content instances?
> * There are quite a few uses of the `"\n"` string literal in a content or pre-content context where the `NL` constant couuld be used.

1. I did wonder about moving the methods and constant into `Content` instead of `Text` but I think it is better to keep `Content` more abstract, and the new methods and constant are more "text-y".  I also like the definitions being up in the `markup` package. 

2. Yes, there may be references in the code to `\n`, but I don't think we need to go overboard to find and change them. To me, this is (generally) another reason to use simple `\n` in text strings, because that is the more usual convention in Java source code.  Note that sometimes these occurrences originate in properties files, and so cannot use the `NL` constant, and it seems silly to use `.replace("\n", NL)` on all such strings.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/markup/RawHtml.java (hunk 3 above)
Comment (hns) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/markup/RawHtml.java:161]: Sorry, I somehow misread the diff.

**Review by hns — APPROVED** [review]
Looks good!

**Pull request closed without merging**